### PR TITLE
Update required members spec

### PR DESCRIPTION
**Pull request by 333fred**
Updating the required members spec with feedback from the most recent 2 LDM meetings on the feature.

### CODE REVIEW

**Comment by 333fred** [comment]
Additionally, I updated the names of the attributes to the names decided at the framework design meeting yesterday.

File: proposals/required-members.md [hunk 1]
```diff
@@ -296,34 +296,38 @@ unrequiring a property with an override in the future, we have design space to d
 Overrides are allowed to mark a member `required` where it was not `required` in the base type. A member so-marked is added to the required members
 list of the derived type.
 
+Types are allowed to override required virtual properties. This means that if the base virtual property has storage, and the dervied type tries to
```
Comment (RikkiGibson) [inline proposals/required-members.md:299]: nit: 'dervied' -> 'derived'

File: proposals/required-members.md [hunk 2]
```diff
@@ -333,72 +337,21 @@ generated. If the constructor already has an `Obsolete` attribute, no additional
...
-base type `Tb` (recursively until reaching `System.Object`) are collected. At every `Ti`, if it has a `RequiredMembersAttribute` with members `R1`..`Rn`,
+**TODO**: LDM has indicated that we desire some kind of dedicated poison attribute, specifically for compilers to indicate that a component is not
+usable in downlevel compilers without resorting to a modreq (ie, a binary breaking change). This will need to be designed and applied to the constructor
+in addition to Obsolete.
+
+To build the full list of `required` members for a given type `T`, including all base types, all the `RequiredMemberAttribute`s from `T` and its
+base type `Tb` (recursively until reaching `System.Object`) are collected. At every `Ti`, if it has a `RequiredMemberAttribute` with members `R1`..`Rn`,
 the following algorithm is performed:
 
 1. Perform standard member lookup on `Ti` for name `Ri` with 0 arguments.
```
Comment (RikkiGibson) [inline proposals/required-members.md:348]: nit: is this really a lookup or better described as simply enumerating the members?

**Review by RikkiGibson — APPROVED** [review]

File: proposals/required-members.md (hunk 2 above)
Comment (333fred) [reply proposals/required-members.md:348]: I meant to update this section entirely, will do so in the morning.

**Comment by 333fred** [comment]
@RikkiGibson please take another look.

**Review by RikkiGibson — APPROVED** [review]

**Comment by 333fred** [comment]
@jcouv please take a look.

File: proposals/required-members.md [hunk 3]
```diff
@@ -296,34 +296,38 @@ unrequiring a property with an override in the future, we have design space to d
...
 ### Metadata Representation
 
 The following 2 attributes are known to the C# compiler and required for this feature to function:
 
 ```cs
 namespace System.Runtime.CompilerServices;
 
-[AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct, AllowMultiple = false)]
-public sealed class RequiredMembersAttribute : Attribute
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Field | AttributeTargets.Property, AllowMultiple = false, Inherited = false)]
```
Comment (jcouv) [inline proposals/required-members.md:310]: `AttributeTargets.Class | AttributeTargets.Struct`: why are we keeping those?

File: proposals/required-members.md [hunk 4]
```diff
@@ -296,34 +296,38 @@ unrequiring a property with an override in the future, we have design space to d
...
+[AttributeUsage(AttributeTargets.Constructor, AllowMultiple = false, Inherited = false)]
+public sealed class SetsRequiredMembersAttribute : Attribute
 {
-    public NoRequiredMembersAttribute() {}
+    public SetsRequiredMembersAttribute() {}
 }
 ```
 
-It is an error to manually apply `RequiredMembersAttribute` to a type.
+It is an error to manually apply `RequiredMemberAttribute` to a type.
```
Comment (jcouv) [inline proposals/required-members.md:323]: How about `SetsRequiredMembersAttribute`?

File: proposals/required-members.md (hunk 3 above)
Comment (jcouv) [reply proposals/required-members.md:310]: Ah, got it. Like extension methods, we want to mark containers.

**Review by RikkiGibson — APPROVED** [review]

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 3)

**Pull request merged**